### ISSUE TITLE
[Discussion] BULK API support in CCR Plugin

### ISSUE THREAD

**Issue by monu-aws** [opensearch-project/cross-cluster-replication#765]
# Background
Most of OpenSearch index level API's support operations on multiple indices. However Cross cluster replication plugin allows users to operate only on one index at a time. 

When 2 clusters are in active passive mode, it becomes very difficult to stop replication running on multiple indices when a fail over is needed. This problem is faced again if the user wants to pause / resume replication on indices during cluster maintenance. To address these limitation, we will be implementing BULK API's which will allow users to control replication on multiple indices with just one API call.

### Functional requirement 

* The API's must cover 3 operations: stop, pause and resume.
* The API should give details on partial success/failures while performing any operation.

### Non-functional requirement

* Maintain security posture of the existing plugin.
* Performance must be significantly better than iterating on individual index with existing API's.
* Keep the cluster healthy throughout and not overload leader/follower
* Cancelling issued operation

### Out of scope ###

* Retry mechanisms on failed indices.

# High level Design

To allow users to control replication on multiple indices without having to list down every index, we will allow users to select indices based on wildcard patterns, this will allow users to quickly stop/pause/resume replication for multiple indices. This method will also allow users to use * as pattern to select all indices.

### Why go with ASYNC workflow

* Asynchronous APIs enable users to submit tasks without waiting for a complete response, reducing the risk of timeouts that can occur when performing operations on a large number of indices.
* An asynchronous API will also provide users with the ability to cancel any submitted replication task.
* Asynchronous API allows users to fetch current status of the replication task. This allows for early detection of failures and STOP if any un-intended action is being taken.

### Why not extend existing replication API’s

* _Breaks_ _existing API contract.[_Different responses] ##TODO
* All the operations (start/stop/pause/resume) rely on index name being present in the URI, allowing customers to pass index pattern in URI will require additional encoding which can result operations being performed on unintended indices. This causes confusion and bad user experience. 
* Different API behavior when a pattern is present vs a index name. (presence of *task_id* when pattern is present)
* All operations replication operations require index name in URI. It is not intuitive when additional filters are passed in request body. Having all index selectors (pattern and filters) allows for better customer experience. 
* It is better to keep existing replication API and bulk API separate as Existing APIs provide response synchronously and the bulk API will only acknowledge the request and complete response will be available only after task is finished.
* Existing API is not cancellable where as new API’s are cancellable.

### Concurrent BULK requests
To avoid any potential conflicts, it is necessary to allow a user to only run one Bulk Replication task at any given time. 

However, user will be able to use existing API’s in tandem with BULK API’s. There will be no additional handling on concurrency between the two API’s.

The Bulk API are not idempotent as in submitting similar request back to back will create another task as it is possible that the old task completed before new task admission.
![concurrency(1)](https://user-images.githubusercontent.com/87794531/229725261-2927de03-63bb-4bca-be36-8efda0c1e607.png)

**STOP request**
``` bash
POST /_plugins/_replication/_bulk_stop?pretty
{
  "pattern" : "fruits-*"    // pattern is mandatory 
}
```

**STOP Response**
```bash
{
  "acknowedged": true,
  "task_id": "<task_id>"   // task_id can be polled to check status of operation
  
}
```

**POLL Request**

```bash
GET /_plugins/_replication/_task_status/<task_id>
```

**POLL Response**
```bash
{
   "operation_type": "stop_replication"              
   "pattern":"fruits-*",
   "num_success": 8,              // indices on which operation was successful           
   "num_failed": 2,               // indices on which operation has failed
   "num_pending": 4,              // indices on which operation is pending
   "start_time": 16294592495
   "failed_indices":[             // Error messages if any 
       {
          "index": "fruits-2",
          "failure_reason": "<Failure reason>"
       },
       {
          "index": "fruits-3",
          "failure_reason": "<Failure reason>"
       }
   ]
}
```

**Cancel BulkTask request**
```bash
POST /_plugins/_replication/_task_cancel/<task_id>
{}
```

**Cancel BulkTask Response**
```bash
{ "acknowedged": true}
```

# Low level design
When a bulk API request is received, a transient task is spawned and the task_id along with a success acknowledgement is returned to the client.
The spawned BulkTask will generate a super list of indices which fall under the provided pattern. Unqualified indices based on filters are filtered out.
Based on a replication setting  plugins.replication.follower.bulk_batch_size , batch the final index list and perform replication action batch by batch.
Customer will be able to  set  plugins.replication.follower.bulk_batch_size  based on their cluster configuration to allow faster completion of bulk task.

### Security Model: Cluster level permissions

Provide cluster level permissions which allows access to BULK API’s.
`cross_cluster_replication_bulk_access`

**Pros**

* Better performance when compared to performing operations one after another. 

* Allows for reduced cluster state updates

* Allows to reduce number of security checks as permission will be checked only once.

* Allows more flexibility in terms of batching operations when executing the bulk task. 
* Most of the customer would be using BULK API to manage cluster level replication. A user when using the role will be assured that replication operation will be performed on all eligible indices and will not fail due to “*insufficient* *permission”*.
* During a failover scenario, user can be notified immediately if they will be able to use BULK API’s. On extreme end, this can save time during failover activity as user does not have to wait for the task to complete only to learn they do not have access to most of the indices. 

**Cons**

* Having cluster level permission will go against fine-grained access control mode. if a user wants to operate using BULK API, they will inherently have permissions over all indices.

### BulkCrossClusterReplicationTask
This is a transient task which will be used to perform Bulk replication and track the progress.
The state of the Bulk task is stored in the cluster state by the master.

**Comment by skumarp7** [comment]
Hi @monusingh-1 , 

any reason for abandoning the PR for bulk API?
